### PR TITLE
HPCC-23140 Refactor/speed up verifyAll and add per connect timeout

### PR DESCRIPTION
**Pull request by jakesmith**
And pass shorter per connection timeout, such that Thor fails
earlier if slaves have not come up in a timely way on startup.

Used to have to wait for watchdog and consequently wait approx.
30 minutes, in some circumstances, if slaves on channels failed
to start up.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/mp/mpcomm.cpp
Comment (jakesmith) [inline system/mp/mpcomm.cpp]: NB: moved from global scope to instance scope.
It protects verifyAll/verifyConnection/disconnect from one another, I can't see why it was at global scope, and my test involved multiple overlapping verifyAll's in fact, I think it was only luck that worked before because of the way that verifyAll backs off for 100ms in a loop.

But please look/consider any possible issues here.

**Comment by jakesmith** [comment]
@mckellyln - please review

**Comment by richardkchapman** [comment]
Typo (conncetion) in commit message

**Comment by jakesmith** [comment]
Fixed typo in commit message.

@mckellyln - please review.

File: system/mp/mpcomm.cpp
Comment (mckellyln) [inline system/mp/mpcomm.cpp]: This routine has a CS in it which may serialize execution.

Comment (jakesmith) [reply system/mp/mpcomm.cpp]: yes, in previous draft of this PR, I had a CS per rank,
but in fact originally the CS was held at the start of verifyAll, with some nasty logic to periodically yield if it failed. That was because the CS was global and multiple communicators (e.g. belonging to different channels) could clash and deadlock each other.
The CS is now inside the communicator and it is correct that it is held for the duration of verifyAll, to exclude other calls to verifyAll or disconnect.
(and that means that the threads no longer block each other)

**Comment by jakesmith** [comment]
@mckellyln resolved issue with badly placed CS.

File: system/mp/mpcomm.cpp
Comment (mckellyln) [reply system/mp/mpcomm.cpp]: I still do not see new threads or any parallelism

Comment (jakesmith) [reply system/mp/mpcomm.cpp]: there's no CS now. So not sure what you mean, can you elaborate?

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: Doesn't the f.get() block ?

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: Should the f.get() be a short wait_for and future_status check is ready ?

Comment (jakesmith) [reply system/mp/mpcomm.cpp]: Shouldn't be necessary, by the time it's in the loop calling f.get() the last N will be running (they will have gained a slot), and will be in the threadedVerifyConnectFunc. They must timeout after approx the overall timeout amount or per connection timeout if shorter.

We need to wait for them to finish either way before leaving verifyAll, so I think f.get() is good.

File: system/mp/mpcomm.cpp
Comment (mckellyln) [inline system/mp/mpcomm.cpp]: Should this be: ?
threads.push_back(std::async(std::launch::async,threadedVerifyConnectFunc, rank, connectFunc));
Without the std::launch::async I don't think new threads are always created ??

File: system/mp/mpcomm.cpp
Comment (mckellyln) [reply system/mp/mpcomm.cpp]: It may be related to std::launch::async ?  See other comment about this.

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: Unrelated, but also if we could eliminate the connectsect CS held for connect() it might offer a potential speedup.

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: not sure, the default behaviour is std::launch::async | std::launch::deferred, which does mean it's implementation dependent, so might limit the number of threads, e.g. if resources are tight.
That may not be a bad thing though.
But I'm not seeing that, I see up to 32 concurrent threads (based on my hardware_concurrency() limit).

Are you not seeing threads created?

I doubt matters much either way here, but given there's a semaphore limiting max here too, it could be changed to std::launch::async.

But I'm curious if you're not seeing thread and why.

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: Correct, no new threads, even with delays added, unless I add std::launch::async.

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: Would it be a compiler version issue ?

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: Or that there are already several threads running and my limit is lower (8) ?  I will try with a forced larger value to see.

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: I tried a larger value (32) and it did not make any difference.

Comment (jakesmith) [reply system/mp/mpcomm.cpp]: I'm not sure why we're seeing very different behviour, I see the max number of threads running in parallel, you see 1.
But as the semantics mean it is guaranteed, and this impl. is self limiting anyway, I will change to std::launch::async

**Comment by jakesmith** [comment]
@mckellyln - please check after change to std::launch::async

File: system/mp/mpcomm.cpp
Comment (mckellyln) [inline system/mp/mpcomm.cpp]: I don't think this works if the group->ordinality() is larger than std::thread::hardware_concurrency() ?

File: system/mp/mpcomm.cpp
Comment (mckellyln) [inline system/mp/mpcomm.cpp]: It sort of seems that the above loop over ranks could be a while and interleave in that loop the check of each thread's completion status and if complete then a get() ?

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: I think its ok as is, but since get() blocks it does seem like we are waiting on them in order.

Comment (jakesmith) [reply system/mp/mpcomm.cpp]: get() on future is effectively a join() and then getting the result from the thread.

The threads will run asynchronously, yes you could be blocked on 1st get() (joining), whilst 2->n had already finished.

But you have to wait on all anyway (have to join all), so there's no value in timing out and joining others instead.
If were waiting blocked (joining) on 1 and others had finished, it would join all others as soon as became unblocked.

I don't think there's a better way of coding it or any gain.
But I could be missing something in what you mean.
@mckellyln  - do you have some pseudo code of what you mean?

Comment (mckellyln) [reply system/mp/mpcomm.cpp]: See my comment above this one.  I think there may be a problem if group->ordinality() > std::thread::hardware_concurrency() ?  And then to change it to a while loop to handle that __suggests__ to me that one might interleave the checking for complete in some threads while still waiting for others.  But yes I agree and stated above, its ok to get() afterwards - as long as we started all we need to.

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: @mckellyln - why? (and I have tested with large groups)

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: >one might interleave the checking for complete in some threads while still waiting for others

still not sure what point would be. But an example may help me understand benefits.

**Comment by richardkchapman** [comment]
Seems to include an unwanted submodule change

**Comment by jakesmith** [comment]
Removed accidental update to submodule.

File: system/mp/mpcomm.cpp
Comment (mckellyln) [reply system/mp/mpcomm.cpp]: ok, I agree I thin it is ok.

File: system/mp/mpcomm.cpp
Comment (mckellyln) [reply system/mp/mpcomm.cpp]: I'll agree it is ok as coded.

**Comment by mckellyln** [comment]
@jakesmith, one last question - do you think as a part of this the connectsect CS held for connect() in verifyConnection() could be looked at ?

**Comment by jakesmith** [comment]
>do you think as a part of this the connectsect CS held for connect() in verifyConnection() could be looked at ?

Possible, but not that related. The connectsect prevents contention dealing with the same CMPChannel.
To avoid that CS, e.g. when multiple threads send to avoid packet clashing etc., is a different problem. And I suspect the CS is reasonable there.

The verifyAll CS (verifysect) is preventing multiple verifyAll's or disconnects clashing on the same communicator (prior to this change, it was preventing any running concurrently, even if different communicators).

I'm not sure looking at the connect() CS usage is worth it, but if you have some further thoughts on it, please open a separate JIRA.

@mckellyln

**Comment by mckellyln** [comment]
@jakesmith I am good with this PR.  ok to merge after squash.

**Review by mckellyln — APPROVED** [review]
good to merge after squash

**Comment by jakesmith** [comment]
@richardkchapman - please merge.

**Comment by richardkchapman** [comment]
@ghalliday Can you sanity-check

File: system/mp/mpcomm.cpp
Comment (ghalliday) [inline system/mp/mpcomm.cpp]: Out of interest, why did you choose to use the standard library functionality, rather than the CAsyncFor?

File: system/mp/mpcomm.cpp
Comment (ghalliday) [inline system/mp/mpcomm.cpp]: Probably not significant in this context, but this will be a vector of fairly large objects.   They will contain synchronization objects.  A quick look at the headers shows them using shared_ptr to shared the state.  They take up about 48 bytes each on my machine.

**Review by ghalliday — COMMENTED** [review]
@jakesmith various questions and comments.

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: I'll change it to "isRankToVerifyFunc"

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: yes, I'll remove

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: CAsyncFor didn't quite fit.

I want to iterate through ranks, adding threads if room in the pool, block if not.
I could do it with a CAsyncFor, but it would immediately launch N threads for each rank, then return immediately for the ones that do nothing, or similar.
Lambda's and closures also helper to capture the variables in use here, which would have not been so clean if using a a derived class for CAsyncFor.

std::future is also a nice paradigm that works well here.

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: ok, I thought because byte it was atomic,
but I didn't realize because it's a local var, that compiler may assume it isn't changed.
Now you point out, I wonder if because it's captured and used in a lambda - the compiler is wise enough to know it might be used/changed by another thread.

Also didn't realize that making it an atomic had no overhead if only reading and writing.
I agree it is clearer as an atomic.

Will change it.

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: will change to getAffinityCpus()

File: system/mp/mpcomm.cpp
Comment (jakesmith) [reply system/mp/mpcomm.cpp]: I may be missing what you're getting at..

If the rank loop that's launching threads (creating std::future<std::async>>'s), stops early,
then this loop will have less future's to iterate through, but all that were created and add to 'threads', will be running at this stage. So these future::get()'s will block until they finish.

**Comment by jakesmith** [comment]
@ghalliday - please see replies and changes.

File: system/mp/mpcomm.cpp
Comment (ghalliday) [reply system/mp/mpcomm.cpp]: Should really be called results rather than threads.

**Review by ghalliday — COMMENTED** [review]
@jakesmith @richardkchapman I think this is fine and ready to merge.  (Up to Jake if he renames the threads variable.)

I have some hesitation about not using CAsyncFor (and possibly extending that with a filter function) since this code is very differently structured to other code we have that is performing async operations.  std::futures are neat, but I would be hesitant to use them in contexts that are time or memory critical - I think it will mean there are more synchronization operations is bound to affect performance.

A general concern (which I have also hit when writing code that uses vectors) is that mixing jlib classes and std classes can be very confusing.  I'm not sure what the solution is, but something we should discuss properly (at the offsite?) rather than drifting into it.

**Comment by jakesmith** [comment]
renamed threads to results, and squashed.

@ghalliday @richardkchapman -ready to merge

**Comment by ghalliday** [comment]
@richardkchapman

**Comment by richardkchapman** [comment]
@jakesmith  Feels a little large for a point release?

**Comment by jakesmith** [comment]
yes I agree, I'll move to master/7.8

**Comment by jakesmith** [comment]
@richardkchapman - retargeted to master.

**Pull request merged**
